server/etcdserver/txn/txn_test.go

```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
+	}{
+		{
+			name: "Unauthorize out of range compare",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("boo"),
+					},
+				},
+				Success: []*pb.RequestOp{},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request range in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: &pb.RangeRequest{
+								Key:      []byte("foo"),
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorize request range out of range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: &pb.RangeRequest{
+								Key:      []byte("boo"),
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request put",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request pur in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: &pb.PutRequest{
+								Key: []byte("foo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorized request pur in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: &pb.PutRequest{
+								Key: []byte("boo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil delete range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestDeleteRange{
+							RequestDeleteRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize delete range in range compare and rerquest",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("foo"),
+					},
+				},
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestDeleteRange{
+							RequestDeleteRange: &pb.DeleteRangeRequest{
+								Key:      []byte("foo"),
+								RangeEnd: []byte("zoo"),
+								PrevKv:   true,
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorize delete range out of range keys",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("foo"),
+					},
+				},
+				Success: []*pb.RequestOp{
```

Comment:
Is there any reason you test only Txn.Success and not Txn.Failure? Either of them can execute depending on compare result.

Reply:
I added the tests